Add package build

labkey-alan opened the pull request:
#### Rationale
This PR adds a shared package build to be used across our various packages (e.g. workflow)

#### Related Pull Requests
* https://github.com/LabKey/sampleManagement/pull/927

#### Changes
* Build: Add shared package build
* Build: Add eln path to app build
* Components: use shared package build
* Components: Move exported test helpers to /internal
* Components: Move un-exported test helpers to /test

labkey-alan committed:
Add package build

labkey-alan committed:
Move app scss files to theme/scss

labkey-alan commented:
At the moment I have the build working for ui-components and workflow with the following issues:

1. The build is not outputting d.ts files, which we definitely need for our packages.
2. Intellij is complaining about all JSX in ui-components because it can't seem to find a tsconfig file. I don't know how to get Intellij to find the tsconfig in our build node_modules directory.

labkey-alan committed:
Remove css/scss from package build, update externals

labkey-alan committed:
Remove scss file imports from ts(x) files

labkey-alan committed:
Configure ui-components to use package build

labkey-alan commented:
I have the d.ts files working now, but still having issues with the tsconfig despite still having one (which extends the one from our build).

labkey-susanh commented:
@labkey-alan Within IntelliJ, we had already set an option for Typescript under Preferences -> Languages & Frameworks -> TypeScript (For me, it's currently `-p /Users/susanhert/Development/labkey/root/server/modules/platform/core/node_modules/@labkey/build/webpack/tsconfig.json`).  

Would something similar help for what you're seeing.

labkey-alan commented:
@labkey-susanh yes that will work, the only other solution as far as I can tell is to have another tsconfig in each package repo. I think your solution is probably the best though.

labkey-alan commented:
On my machine that flag only works if I use a relative path from project root. So for my ui-components project I have to set it as such: `-p packages/build/webpack/tsconfig.json`.

labkey-alan commented:
Ok, at this point I think I have the build working properly, but I cannot get the tests to run. Without a tsconfig the tests throw errors during configuration in jest.setup.ts, if I pass a tsconfig path to ts-jest via `jest.globals` in the `package.json` any test referencing the LABKEY object throws an error because LABKEY cannot be found.

labkey-alan commented:
I have tried a pile of different things to get the ui-components tests to work with a tsconfig provided by our build:

1. Moved our jest config from package.json to jest.config.js
   * My thought here was that maybe something was transforming the value in the package.json and that if it's a JS config that transformation wouldn't happen
2. Added a tsconfig.json that only contains `"extends": "./node_modules/@labkey/build/tsconfig.json"`
3. Changed the tsconfig in the build package to match exactly what we had in the ui-components tsconfig (there was some slight differences)

None of these changes work. Not individually, not together. They all result in our tests complaining that they cannot find the name 'LABKEY':

```
 FAIL  src/internal/components/domainproperties/models.spec.ts
  ● Test suite failed to run

    src/internal/components/domainproperties/models.spec.ts:240:5 - error TS2304: Cannot find name 'LABKEY'.
```

Currently the only known fix is to put a tsconfig in the components directory. I am assuming this is true for all of our other packages as well.

labkey-alan commented:
I have tried all the steps in the first half of the ts-jest troubleshooting guide as well:

https://github.com/kulshekhar/ts-jest/blob/main/TROUBLESHOOTING.md

At this point I am thinking it may be best to just fix these test files, they shouldn't be directly touching the `LABKEY` object anyway. Before I go down this path does anyone else want to take a crack at it? @labkey-nicka @cnathe @labkey-susanh ?

labkey-alan commented:
Also, without a tsconfig in the components directory Intellij cannot find jest globals.

<img width="715" alt="Screen Shot 2022-04-25 at 11 24 04 AM" src="https://user-images.githubusercontent.com/5341647/165131901-d04e566a-7f9a-45ab-8a94-6ea3f690f5f2.png">

labkey-alan committed:
Move jest config from package.json to jest.config.js

labkey-alan committed:
Only add outDir during package build, update tsconfig.json

labkey-alan committed:
Update components tsconfig.json so tests work

labkey-alan committed:
Rearrange exported/non-exported test files

All exported test code lives in /internal
Any non-exported test code lives in /test

labkey-alan committed:
Bump build, update build release notes

labkey-alan committed:
Update components release notes

labkey-alan committed:
Bump components

labkey-alan committed:
Add @labkey/components to externals in package build

labkey-alan committed:
Bump components

labkey-alan committed:
QueryInfo: declare appEditableTable

labkey-alan committed:
hasModule: add optional moduleContext arg

Needed or tests cannot access moduleContext set via mountWithServerContext

labkey-alan committed:
Bump components

labkey-alan committed:
Remove commented out code

labkey-alan committed:
tsconfig.json: remove es2015 lib

labkey-alan committed:
Bump build

labkey-alan committed:
Make include directives more specific to build types

labkey-alan committed:
Bump build

labkey-alan committed:
Update tsconfig.json

labkey-alan committed:
Correct include/exclude directives

labkey-alan committed:
Update components tsconfig.json to add typings as type root

labkey-alan committed:
Bump build and components

labkey-alan committed:
Bump components

labkey-alan committed:
Bump components

labkey-alan committed:
Build tsconfig.json: re-enable sourceMap

labkey-alan committed:
Prep for release

labkey-alan committed:
Fix ResponsiveMenuButtonGroup.spec.tsx

labkey-alan committed:
Fix SamplesAddButton.spec.tsx, SamplesAssayButton.spec.tsx
